### DISCUSSION TITLE
Bug with "is not Class" disambiguation

### DISCUSSION THREAD

**Discussion by Joy-less** [dotnet/csharplang#7980] [Language Ideas]
There's an annoying double standard with C#'s `is` disambiguation:

```cs
public class Program {
    public class Cat {
    }
    public static void Main() {
        Cat Cat = new();
        object Object = "test";
        
        // Compiles fine
        if (Object is Cat) {
            
        }
        
        // Compilation error (line 14, col 21): A constant value is expected
        if (Object is not Cat) {
            
        }
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Yes.  THis is a known deviation due to how patterns work.  `is Cat` is hte prior existing lang feature, which has to pick the `is type` semantics due to back compat. 

But `is not Pattern` allows for any `Pattern` after the `not` piece.  Here `Pattern` could be referencing a constant value (as you can say `is not constant`, or could be referencing a type.  In this case, because `Cat` could reference either (it's legal as a type name lookup or an expression name lookup), we consider it ambiguous.  

This is definitely a tripping point and an unfortunate case where keeping legacy behavior for an existing form is necessary, but it ends up being inconsistent with the new syntax.

If we had patterns from v1, these would be consistent.

**Reply by alrz** [reply]
This is not the first time that this came up. Would it be worthwhile to take a breaking change for this sort of papercuts, now that there's a clear path to do that? I think this is not far off the criteria laid out at https://github.com/dotnet/csharplang/issues/7918

**Reply by CyrusNajmabadi** [reply]
If someone is willing to champion and it gets put on the schedule.  To me, it's not worth it. But maybe another ldm member feels differently.

**Reply by alrz** [reply]
Just to mention, there's also keywords or, not, and, that can be bound as either a variable OR a type, for example 'x is or not' is completely valid (refer to https://github.com/dotnet/roslyn/pull/49621 to see the gymnastics done to keep this behavior while fixing a bug), but I'd agree that these are probably not worth changing except if https://github.com/dotnet/csharplang/issues/4460 was to be fully embraced. (I don't think the new proposal was meant to follow that direction though)

**Reply by Joy-less** [reply]
If you type this:
```csharp
if (!(Object is Cat)) {

}
```
You get a suggestion (`IDE0083: Use Pattern Matching`) which corrects your code to the compilation error code.

**Reply by CyrusNajmabadi** [reply]
> If you type this:
> ```csharp
> if (!(Object is Cat)) {
> 
> }
> ```
> You get a suggestion (`IDE0083: Use Pattern Matching`) which corrects your code to the compilation error code.

@Joy-less such issues can be filed as bugs at dotnet/Roslyn.  They are not part of the language design. Thanks!
